File: scripts/auto-capture.py
#!/usr/bin/env python3
"""
Smart IQ Capture - Signal-triggered recording with pre/post buffers.

Monitors rtlmux/rtl_tcp stream and captures IQ when signal is detected.
Uses a state machine with hysteresis to capture full transmissions.

Features:
- Pre-trigger buffer to capture signal preamble
- Post-trigger tail to capture signal end
- Separate trigger/release thresholds (hysteresis)
- Smoothed std to prevent false triggers
- Stats file for each capture
- Quiet mode for minimal output

Usage:
    python3 auto-capture.py --host 192.168.1.69 --port 1235 --threshold 1.5 --output /data/debug_audio
    python3 auto-capture.py --threshold 1.5 --release 1.2 --pre 2.0 --post 1.0 --output /data/debug_audio
    python3 auto-capture.py --quiet --max 20 --timeout 3600 --output /data/debug_audio
"""

import socket
import struct
import time
import argparse
import sys
from collections import deque
from pathlib import Path
from datetime import datetime

import numpy as np


def _running_in_container() -> bool:
    if Path("/.dockerenv").exists():
        return True
    try:
        cgroup = Path("/proc/1/cgroup").read_text(errors="ignore")
        return "docker" in cgroup or "containerd" in cgroup
    except Exception:
        return False


class SignalCapture:
    """Smart IQ capture with state machine and hysteresis."""
    
    HEADER_SIZE = 12
    SAMPLE_RATE = 2_048_000  # 2.048 Msps (as configured in rtl_tcp)
    
    def __init__(
        self,
        host: str = "192.168.1.69",
        port: int = 1235,
        trigger_threshold: float = 1.5,
        release_threshold: float | None = None,
        release_hold_sec: float = 0.15,
        pre_buffer_sec: float = 1.0,
        post_signal_sec: float = 0.5,
        min_duration_sec: float = 0.5,
        max_duration_sec: float = 60.0,
        cooldown_sec: float = 2.0,
        output_dir: str = "/data/debug_audio",
        quiet: bool = False,
    ):
        self.host = host
        self.port = port
        self.trigger_threshold = trigger_threshold
        self.release_threshold = release_threshold or (trigger_threshold * 0.85)
        self.release_hold_sec = release_hold_sec
        self.pre_buffer_sec = pre_buffer_sec
        self.post_signal_sec = post_signal_sec
        self.min_duration_sec = min_duration_sec
        self.max_duration_sec = max_duration_sec
        self.cooldown_sec = cooldown_sec
        self.output_dir = Path(output_dir)
        self.quiet = quiet
        
        # Derived values
        self.bytes_per_sec = self.SAMPLE_RATE * 2
        self.chunk_size = 32768  # ~8ms at 2.048 Msps
        
        # Ring buffer for pre-trigger
        pre_buffer_bytes = int(self.pre_buffer_sec * self.bytes_per_sec)
        self.pre_buffer_chunks = pre_buffer_bytes // self.chunk_size + 1
        
        self.output_dir.mkdir(parents=True, exist_ok=True)
    
    def _log(self, msg: str, end: str = "\n", flush: bool = False):
        """Print message if not in quiet mode."""
        if not self.quiet:
            print(msg, end=end, flush=flush)
    
    def _progress(self, msg: str):
        """Print progress line (overwritten)."""
        if not self.quiet:
            sys.stdout.write(f"\r{msg}   ")
            sys.stdout.flush()
    
    def connect(self) -> socket.socket:
        """Connect to rtlmux/rtl_tcp and read header."""
        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        sock.settimeout(30.0)
        self._log(f"Connecting to {self.host}:{self.port}...")
        sock.connect((self.host, self.port))
        
        # Read RTL-TCP header
        header = sock.recv(self.HEADER_SIZE)
        if len(header) < self.HEADER_SIZE:
            raise RuntimeError("Failed to receive RTL-TCP header")
        
        magic = header[:4]
        tuner_type = struct.unpack(">I", header[4:8])[0]
        self._log(f"Connected! Magic: {magic}, Tuner type: {tuner_type}")
        
        sock.settimeout(1.0)
        return sock
    
    def analyze_chunk(self, data: bytes) -> float:
        """Calculate std dev of chunk."""
        arr = np.frombuffer(data, dtype=np.uint8)
        return float(np.std(arr))
    
    def save_capture(self, data: bytes, stats: dict) -> Path:
        """Save captured IQ data and stats file."""
        timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
        iq_file = self.output_dir / f"iq_capture_{timestamp}.u8"
        stats_file = self.output_dir / f"iq_capture_{timestamp}_stats.txt"
        
        # Save IQ data
        with open(iq_file, "wb") as f:
            f.write(data)
        
        # Save stats
        duration = len(data) / self.bytes_per_sec
        with open(stats_file, "w") as f:
            f.write(f"Capture: {iq_file.name}\n")
            f.write(f"Timestamp: {timestamp}\n")
            f.write(f"Size: {len(data):,} bytes\n")
            f.write(f"Duration: {duration:.2f}s\n")
            f.write(f"Sample rate: {self.SAMPLE_RATE:,} Sps\n")
            f.write(f"Peak std: {stats.get('peak_std', 0):.2f}\n")
            f.write(f"Mean std: {stats.get('mean_std', 0):.2f}\n")
            f.write(f"Trigger threshold: {self.trigger_threshold}\n")
            f.write(f"Release threshold: {self.release_threshold}\n")
        
        return iq_file
    
    def monitor(self, max_captures: int = 10, timeout_sec: float = 300) -> int:
        """Monitor stream and capture on signal detection."""
        sock = self.connect()
        
        captures = 0
        start_time = time.time()
        
        # State machine
        state = "WAITING"  # WAITING, CAPTURING, POST_SIGNAL
        ring_buffer: deque[bytes] = deque(maxlen=self.pre_buffer_chunks)
        capture_buffer: list[bytes] = []
        signal_start_time = 0.0
        last_signal_time = 0.0
        last_capture_time = 0.0
        capture_stats = {"peak_std": 0.0, "std_samples": []}

        # Debounce signal end detection (prevents rapid end/resume flapping)
        release_candidate_since: float | None = None
        
        # Smoothing for std (prevent false triggers)
        std_history: deque[float] = deque(maxlen=5)
        
        self._log(f"\n{'=' * 60}")
        self._log(f"Monitoring for signals")
        self._log(f"  Trigger: std > {self.trigger_threshold}")
        self._log(f"  Release: std < {self.release_threshold}")
        self._log(f"  Pre-buffer: {self.pre_buffer_sec}s, Post-tail: {self.post_signal_sec}s")
        self._log(f"  Duration: {self.min_duration_sec}s - {self.max_duration_sec}s")
        self._log(f"{'=' * 60}\n")
        
        try:
            while captures < max_captures:
                # Check timeout
                elapsed = time.time() - start_time
                if timeout_sec > 0 and elapsed > timeout_sec:
                    self._log(f"\nTimeout reached ({timeout_sec}s)")
                    break
                
                # Read chunk
                try:
                    data = sock.recv(self.chunk_size)
                    if not data:
                        self._log("Connection closed")
                        break
                except socket.timeout:
                    continue
                
                # Analyze
                std = self.analyze_chunk(data)
                std_history.append(std)
                smoothed_std = float(np.mean(std_history))
                
                now = time.time()
                
                if state == "WAITING":
                    ring_buffer.append(data)
                    
                    # Check for trigger (with cooldown)
                    if smoothed_std > self.trigger_threshold and (now - last_capture_time) > self.cooldown_sec:
                        self._log(f"\n🔊 TRIGGERED! std={smoothed_std:.2f}")
                        state = "CAPTURING"
                        signal_start_time = now
                        last_signal_time = now
                        
                        # Start with pre-buffer
                        capture_buffer = list(ring_buffer)
                        capture_buffer.append(data)
                        capture_stats = {"peak_std": smoothed_std, "std_samples": [smoothed_std]}
                        
                        pre_sec = len(b"".join(capture_buffer)) / self.bytes_per_sec
                        self._log(f"  Pre-buffer: {pre_sec:.2f}s")
                    else:
                        # Show progress
                        bar_len = int(min(smoothed_std / 2.0, 1.0) * 30)
                        bar = "█" * bar_len + "░" * (30 - bar_len)
                        marker_pos = min(int(self.trigger_threshold / 2.0 * 30), 29)
                        bar = bar[:marker_pos] + "|" + bar[marker_pos + 1:]
                        self._progress(f"[{bar}] std={smoothed_std:.2f} cap={captures}/{max_captures}")
                
                elif state == "CAPTURING":
                    capture_buffer.append(data)
                    capture_stats["std_samples"].append(smoothed_std)
                    if smoothed_std > capture_stats["peak_std"]:
                        capture_stats["peak_std"] = smoothed_std
                    
                    duration = now - signal_start_time
                    
                    # Check for signal end
                    if smoothed_std < self.release_threshold:
                        if release_candidate_since is None:
                            release_candidate_since = now
                        elif (now - release_candidate_since) >= self.release_hold_sec:
                            self._log(f"\n  Signal ended at {duration:.2f}s")
                            state = "POST_SIGNAL"
                            last_signal_time = now
                    elif duration >= self.max_duration_sec:
                        self._log(f"\n  Max duration reached")
                        state = "SAVING"
                    else:
                        release_candidate_since = None
                        self._progress(f"  Capturing: {duration:.1f}s std={smoothed_std:.2f}")
                
                elif state == "POST_SIGNAL":
                    capture_buffer.append(data)
                    
                    if smoothed_std > self.trigger_threshold:
                        # Signal came back strongly enough to re-trigger
                        state = "CAPTURING"
                        release_candidate_since = None
                        self._log("  Signal resumed")
                    elif (now - last_signal_time) >= self.post_signal_sec:
                        state = "SAVING"
                
                if state == "SAVING":
                    total_duration = now - signal_start_time
                    
                    if total_duration >= self.min_duration_sec:
                        all_data = b"".join(capture_buffer)
                        capture_stats["mean_std"] = float(np.mean(capture_stats["std_samples"]))
                        del capture_stats["std_samples"]
                        
                        filepath = self.save_capture(all_data, capture_stats)
                        captures += 1
                        last_capture_time = now
                        
                        self._log(f"✅ Saved: {filepath.name}")
                        self._log(f"   Duration: {total_duration:.2f}s, Size: {len(all_data):,} bytes")
                        self._log(f"   Peak std: {capture_stats['peak_std']:.2f}, Mean std: {capture_stats['mean_std']:.2f}")
                        self._log("")
                    else:
                        self._log(f"⚠️  Too short ({total_duration:.2f}s < {self.min_duration_sec}s), discarding\n")
                    
                    # Reset
                    state = "WAITING"
                    capture_buffer = []
                    ring_buffer.clear()
                    capture_stats = {"peak_std": 0.0, "std_samples": []}
                    release_candidate_since = None
        
        except KeyboardInterrupt:
            self._log("\n\nInterrupted by user")
            
            # Save in-progress capture
            if capture_buffer and len(b"".join(capture_buffer)) > self.min_duration_sec * self.bytes_per_sec:
                all_data = b"".join(capture_buffer)
                capture_stats["mean_std"] = float(np.mean(capture_stats.get("std_samples", [0])))
                if "std_samples" in capture_stats:
                    del capture_stats["std_samples"]
                filepath = self.save_capture(all_data, capture_stats)
                captures += 1
                self._log(f"Saved partial: {filepath.name}")
        finally:
            sock.close()
        
        self._log(f"\nTotal captures: {captures}")
        return captures


def main():
    parser = argparse.ArgumentParser(
        description="Smart IQ Capture - Signal-triggered recording",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
Examples:
  %(prog)s --threshold 1.5              # Basic monitoring
  %(prog)s --threshold 1.5 --release 1.2 --pre 2.0  # Custom hysteresis
  %(prog)s --quiet --max 20 --timeout 3600  # Long unattended capture
        """
    )
    parser.add_argument("--host", default="192.168.1.69", help="rtlmux/rtl_tcp host")
    parser.add_argument("--port", type=int, default=1235, help="rtlmux/rtl_tcp port")
    parser.add_argument("--threshold", type=float, default=1.5, 
                        help="Trigger threshold (std dev)")
    parser.add_argument("--release", type=float, default=None,
                        help="Release threshold (default: 85%% of trigger)")
    parser.add_argument(
        "--release-hold",
        type=float,
        default=0.15,
        help="Seconds below release threshold required to end signal (debounce)",
    )
    parser.add_argument("--pre", type=float, default=1.0,
                        help="Pre-trigger buffer seconds")
    parser.add_argument("--post", type=float, default=0.5,
                        help="Post-signal tail seconds")
    parser.add_argument("--min-duration", type=float, default=0.5,
                        help="Minimum capture duration")
    parser.add_argument("--max-duration", type=float, default=60.0,
                        help="Maximum capture duration")
    parser.add_argument("--cooldown", type=float, default=2.0,
                        help="Cooldown between captures")
    parser.add_argument("--max", type=int, default=10,
                        help="Maximum number of captures")
    parser.add_argument("--timeout", type=float, default=300,
                        help="Monitoring timeout (0 = no timeout)")
    parser.add_argument("--output", default="/data/debug_audio",
                        help="Output directory")
    parser.add_argument("--quiet", "-q", action="store_true",
                        help="Minimal output")
    parser.add_argument(
        "--allow-host",
        action="store_true",
        help="Allow running outside the demod-test container (advanced)",
    )
    
    args = parser.parse_args()

    if not args.allow_host and not _running_in_container():
        print(
            "This tool is intended to run inside the demod-test container.\n\n"
            "Run it via:\n"
            "  docker compose -f docker-compose.demod-test.yml run --rm demod-test "
            "python3 /scripts/auto-capture.py --output /data/debug_audio\n\n"
            "Or use the interactive wrapper:\n"
            "  node scripts/auto-capture.mjs\n",
            file=sys.stderr,
        )
        raise SystemExit(2)
    
    capture = SignalCapture(
        host=args.host,
        port=args.port,
        trigger_threshold=args.threshold,
        release_threshold=args.release,
        release_hold_sec=args.release_hold,
        pre_buffer_sec=args.pre,
        post_signal_sec=args.post,
        min_duration_sec=args.min_duration,
        max_duration_sec=args.max_duration,
        cooldown_sec=args.cooldown,
        output_dir=args.output,
        quiet=args.quiet,
    )
    
    capture.monitor(max_captures=args.max, timeout_sec=args.timeout)


if __name__ == "__main__":
    main()

File: scripts/auto-capture.mjs
#!/usr/bin/env node

/**
 * Auto Capture (Interactive Wrapper)
 *
 * Always runs IQ auto-capture inside the demod-test container and writes
 * captures to the shared debug_audio volume.
 *
 * Usage:
 *   node scripts/auto-capture.mjs
 *   node scripts/auto-capture.mjs --host 192.168.1.69 --port 1235 --threshold 1.5
 *
 * Options are passed through to /scripts/auto-capture.py.
 */

import { spawn } from "node:child_process"
import { access, mkdir } from "node:fs/promises"
import { EOL } from "node:os"
import { resolve } from "node:path"
import { stdin, stdout, stderr } from "node:process"
import { parseArgs } from "node:util"
import readline from "node:readline/promises"

const IS_TTY = Boolean(stdout.isTTY && stdin.isTTY)

function color(code, text) {
	if (!IS_TTY) return text
	return `\u001b[${code}m${text}\u001b[0m`
}

function dim(text) {
	return color("2", text)
}

function bold(text) {
	return color("1", text)
}

function red(text) {
	return color("31", text)
}

function green(text) {
	return color("32", text)
}

function yellow(text) {
	return color("33", text)
}

function cyan(text) {
	return color("36", text)
}

function clearScreen() {
	if (!IS_TTY) return
	stdout.write("\u001b[2J\u001b[H")
}

function hr(width = 72) {
	return "-".repeat(Math.max(10, width))
}

function clampNumber(value, { min = -Infinity, max = Infinity } = {}) {
	if (Number.isNaN(value)) return value
	return Math.min(max, Math.max(min, value))
}

function parseNumberOrNull(text) {
	const trimmed = String(text ?? "").trim()
	if (!trimmed) return null
	const num = Number(trimmed)
	return Number.isFinite(num) ? num : null
}

async function ensureHostDebugAudioDir() {
	const dir = resolve("debug_audio")
	try {
		await access(dir)
		return dir
	} catch {
		await mkdir(dir, { recursive: true })
		return dir
	}
}

function buildPythonArgs(config) {
	/** @type {string[]} */
	const args = ["python3", "/scripts/auto-capture.py"]

	args.push("--host", config.host)
	args.push("--port", String(config.port))
	args.push("--threshold", String(config.threshold))
	if (config.release != null) args.push("--release", String(config.release))
	args.push("--pre", String(config.pre))
	args.push("--post", String(config.post))
	args.push("--min-duration", String(config.minDuration))
	args.push("--max-duration", String(config.maxDuration))
	args.push("--cooldown", String(config.cooldown))
	args.push("--max", String(config.maxCaptures))
	args.push("--timeout", String(config.timeout))
	args.push("--output", "/data/debug_audio")
	if (config.quiet) args.push("--quiet")

	return args
}

function buildDockerComposeArgs({ composeFile, service, pythonArgs }) {
	return ["compose", "-f", composeFile, "run", "--rm", service, ...pythonArgs]
}

function printConfig(config, { composeFile, service }) {
	stdout.write(
		[
			bold("WaveKit Auto Capture"),
			dim("(runs inside demod-test container; outputs to ./debug_audio)"),
			hr(),
			`${bold("Docker")}: docker compose -f ${composeFile} run --rm ${service}`,
			`${bold("Output")}: /data/debug_audio  ${dim("(host: ./debug_audio)")}`,
			"",
			`${bold("Source")}: ${config.host}:${config.port}`,
			`${bold("Trigger")}: std > ${config.threshold}`,
			`${bold("Release")}: ${config.release ?? dim("(auto: 85% of trigger)")}`,
			`${bold("Buffers")}: pre=${config.pre}s post=${config.post}s`,
			`${bold("Duration")}: min=${config.minDuration}s max=${config.maxDuration}s`,
			`${bold("Limits")}: maxCaptures=${config.maxCaptures} timeout=${config.timeout}s cooldown=${config.cooldown}s`,
			`${bold("Mode")}: ${config.quiet ? "quiet" : "normal"}`,
			hr(),
			"",
		].join(EOL),
	)
}

async function promptText(rl, label, { defaultValue } = {}) {
	const suffix =
		defaultValue != null ? ` ${dim(`(default: ${defaultValue})`)}` : ""
	const answer = await rl.question(`${cyan(label)}${suffix}: `)
	const trimmed = answer.trim()
	if (trimmed) return trimmed
	return defaultValue ?? ""
}

async function promptNumber(rl, label, { defaultValue, min, max } = {}) {
	while (true) {
		const answer = await promptText(rl, label, {
			defaultValue: String(defaultValue),
		})
		const n = parseNumberOrNull(answer)
		if (n == null) {
			stdout.write(red("Please enter a valid number.") + EOL)
			continue
		}
		if (min != null && n < min) {
			stdout.write(red(`Value must be >= ${min}.`) + EOL)
			continue
		}
		if (max != null && n > max) {
			stdout.write(red(`Value must be <= ${max}.`) + EOL)
			continue
		}
		return n
	}
}

async function promptYesNo(rl, label, { defaultValue = false } = {}) {
	const hint = defaultValue ? "Y/n" : "y/N"
	while (true) {
		const answer = await rl.question(`${cyan(label)} ${dim(`(${hint})`)}: `)
		const trimmed = answer.trim().toLowerCase()
		if (!trimmed) return defaultValue
		if (["y", "yes"].includes(trimmed)) return true
		if (["n", "no"].includes(trimmed)) return false
		stdout.write(red("Please answer y or n.") + EOL)
	}
}

async function resolveConfigFromArgs(argv) {
	const defaults = {
		host: "192.168.1.69",
		port: 1235,
		threshold: 1.5,
		release: null,
		pre: 1.0,
		post: 0.5,
		minDuration: 0.5,
		maxDuration: 60.0,
		cooldown: 2.0,
		maxCaptures: 10,
		timeout: 300,
		quiet: false,
	}

	const { values } = parseArgs({
		args: argv,
		allowPositionals: false,
		options: {
			host: { type: "string" },
			port: { type: "string" },
			threshold: { type: "string" },
			release: { type: "string" },
			pre: { type: "string" },
			post: { type: "string" },
			"min-duration": { type: "string" },
			"max-duration": { type: "string" },
			cooldown: { type: "string" },
			max: { type: "string" },
			timeout: { type: "string" },
			quiet: { type: "boolean" },
			interactive: { type: "boolean" },
			"dry-run": { type: "boolean" },
			"compose-file": { type: "string" },
			service: { type: "string" },
			help: { type: "boolean" },
		},
	})

	if (values.help) {
		stdout.write(
			[
				bold("WaveKit Auto Capture (interactive)"),
				"",
				"Runs /scripts/auto-capture.py inside the demod-test container and writes to ./debug_audio.",
				"",
				bold("Usage"),
				"  node scripts/auto-capture.mjs [options]",
				"",
				bold("Options"),
				"  --host <ip>            rtl_tcp host",
				"  --port <n>             rtl_tcp port",
				"  --threshold <n>        trigger threshold (std)",
				"  --release <n>          release threshold (std)",
				"  --pre <sec>            pre-buffer seconds",
				"  --post <sec>           post-tail seconds",
				"  --min-duration <sec>   minimum capture duration",
				"  --max-duration <sec>   maximum capture duration",
				"  --cooldown <sec>       cooldown between captures",
				"  --max <n>              max captures",
				"  --timeout <sec>        timeout (0 = no timeout)",
				"  --quiet                minimal output",
				"  --interactive          force prompts",
				"  --dry-run              print docker command, don’t run",
				"  --compose-file <path>  compose file (default: docker-compose.demod-test.yml)",
				"  --service <name>       service name (default: demod-test)",
			].join(EOL) + EOL,
		)
		process.exit(0)
	}

	const composeFile = values["compose-file"] ?? "docker-compose.demod-test.yml"
	const service = values.service ?? "demod-test"
	const forceInteractive = Boolean(values.interactive)
	const dryRun = Boolean(values["dry-run"])

	const parsed = {
		host: values.host ?? defaults.host,
		port: parseNumberOrNull(values.port) ?? defaults.port,
		threshold: parseNumberOrNull(values.threshold) ?? defaults.threshold,
		release:
			values.release != null
				? parseNumberOrNull(values.release)
				: defaults.release,
		pre: parseNumberOrNull(values.pre) ?? defaults.pre,
		post: parseNumberOrNull(values.post) ?? defaults.post,
		minDuration:
			parseNumberOrNull(values["min-duration"]) ?? defaults.minDuration,
		maxDuration:
			parseNumberOrNull(values["max-duration"]) ?? defaults.maxDuration,
		cooldown: parseNumberOrNull(values.cooldown) ?? defaults.cooldown,
		maxCaptures: Math.round(
			parseNumberOrNull(values.max) ?? defaults.maxCaptures,
		),
		timeout: parseNumberOrNull(values.timeout) ?? defaults.timeout,
		quiet: Boolean(values.quiet ?? defaults.quiet),
	}

	// If any values are missing due to invalid parsing, fall back to defaults.
	parsed.port = Number.isFinite(parsed.port)
		? clampNumber(parsed.port, { min: 1, max: 65535 })
		: defaults.port
	parsed.threshold = Number.isFinite(parsed.threshold)
		? Math.max(0.001, parsed.threshold)
		: defaults.threshold
	parsed.pre = Number.isFinite(parsed.pre)
		? Math.max(0, parsed.pre)
		: defaults.pre
	parsed.post = Number.isFinite(parsed.post)
		? Math.max(0, parsed.post)
		: defaults.post
	parsed.minDuration = Number.isFinite(parsed.minDuration)
		? Math.max(0, parsed.minDuration)
		: defaults.minDuration
	parsed.maxDuration = Number.isFinite(parsed.maxDuration)
		? Math.max(parsed.minDuration, parsed.maxDuration)
		: defaults.maxDuration
	parsed.cooldown = Number.isFinite(parsed.cooldown)
		? Math.max(0, parsed.cooldown)
		: defaults.cooldown
	parsed.maxCaptures = Number.isFinite(parsed.maxCaptures)
		? Math.max(1, parsed.maxCaptures)
		: defaults.maxCaptures
	parsed.timeout = Number.isFinite(parsed.timeout)
		? Math.max(0, parsed.timeout)
		: defaults.timeout
	if (parsed.release != null && !Number.isFinite(parsed.release))
		parsed.release = null

	const needsPrompts = forceInteractive || (IS_TTY && argv.length === 0)
	if (!needsPrompts) return { config: parsed, composeFile, service, dryRun }

	clearScreen()
	stdout.write(bold("WaveKit Auto Capture") + EOL)
	stdout.write(dim("Interactive mode — press Enter to accept defaults") + EOL)
	stdout.write(hr() + EOL + EOL)

	const rl = readline.createInterface({ input: stdin, output: stdout })
	try {
		parsed.host = await promptText(rl, "rtlmux/rtl_tcp host", {
			defaultValue: parsed.host,
		})
		parsed.port = await promptNumber(rl, "rtlmux/rtl_tcp port", {
			defaultValue: parsed.port,
			min: 1,
			max: 65535,
		})
		parsed.threshold = await promptNumber(rl, "Trigger threshold (std)", {
			defaultValue: parsed.threshold,
			min: 0.001,
		})

		const releaseText = await promptText(
			rl,
			"Release threshold (std) — leave blank for auto",
			{
				defaultValue: "",
			},
		)
		parsed.release = releaseText ? parseNumberOrNull(releaseText) : null

		parsed.pre = await promptNumber(rl, "Pre-buffer seconds", {
			defaultValue: parsed.pre,
			min: 0,
		})
		parsed.post = await promptNumber(rl, "Post-tail seconds", {
			defaultValue: parsed.post,
			min: 0,
		})
		parsed.minDuration = await promptNumber(rl, "Min duration seconds", {
			defaultValue: parsed.minDuration,
			min: 0,
		})
		parsed.maxDuration = await promptNumber(rl, "Max duration seconds", {
			defaultValue: parsed.maxDuration,
			min: parsed.minDuration,
		})
		parsed.cooldown = await promptNumber(rl, "Cooldown seconds", {
			defaultValue: parsed.cooldown,
			min: 0,
		})
		parsed.maxCaptures = await promptNumber(rl, "Max captures", {
			defaultValue: parsed.maxCaptures,
			min: 1,
			max: 100000,
		})
		parsed.timeout = await promptNumber(
			rl,
			"Timeout seconds (0 = no timeout)",
			{
				defaultValue: parsed.timeout,
				min: 0,
			},
		)
		parsed.quiet = await promptYesNo(rl, "Quiet mode", {
			defaultValue: parsed.quiet,
		})

		stdout.write(EOL)
		const ok = await promptYesNo(rl, "Start capture now?", {
			defaultValue: true,
		})
		if (!ok) process.exit(0)
	} finally {
		rl.close()
	}

	return { config: parsed, composeFile, service, dryRun }
}

async function main() {
	await ensureHostDebugAudioDir()

	const argv = process.argv.slice(2)
	const { config, composeFile, service, dryRun } =
		await resolveConfigFromArgs(argv)
	const pythonArgs = buildPythonArgs(config)
	const dockerArgs = buildDockerComposeArgs({
		composeFile,
		service,
		pythonArgs,
	})

	printConfig(config, { composeFile, service })

	if (dryRun) {
		stdout.write(yellow("Dry-run:" + EOL))
		stdout.write(
			`docker ${dockerArgs.map(a => JSON.stringify(a)).join(" ")}${EOL}`,
		)
		process.exit(0)
	}

	stdout.write(green("Starting capture…") + EOL)
	const child = spawn("docker", dockerArgs, {
		stdio: "inherit",
		env: process.env,
	})

	child.on("exit", code => {
		if (typeof code === "number" && code !== 0) {
			stderr.write(red(`auto-capture failed (exit ${code})`) + EOL)
			process.exit(code)
		}
		stdout.write(EOL)
		stdout.write(green("Done.") + EOL)
		stdout.write(
			[
				"Next:",
				`  docker compose -f ${composeFile} run --rm ${service} bash`,
				`  # inside: bash /scripts/demod-test.sh /data/debug_audio/iq_capture_*.u8`,
			].join(EOL) + EOL,
		)
		process.exit(0)
	})
}

main().catch(err => {
	stderr.write(red(String(err?.stack ?? err)) + EOL)
	process.exit(1)
})
